### PR TITLE
[Change] Use new Engage Endpoint for engage app version creation

### LINKED ISSUE #216: Use new Engage apps and app-version endpoints
## Todo:
Add the following commands to create and delete apps:
```
deepo platform engage-app create -n NAME  
deepo platform engage-app delete -i ID
```

Add the following commands to create app-versions:
```
deepo platform engage-app-version create [-w WORKFLOW] [-c CUSTOM_NODES] -a APP_ID -r [RECOGNITION_VERSION_IDS [RECOGNITION_VERSION_IDS ...]]  [-p PREVIOUS_ENGAGE_APP_VERSION_ID] 
```
## How:
Use the new Engage endpoints : 
```
POST: /engage/fs/on-site/orgs/${orgName}/apps/
DELETE: /engage/fs/on-site/orgs/${orgName}/apps/${app_id}
POST: /engage/fs/on-site/orgs/${orgName}/apps/${app_id}/versions
```

The Urls are on Vesta Stag

## Note
To implement the create apps:
* Cleanup the previous deepomatic/cli/cmds/platform/engage_app/create.py and deepomatic/cli/cmds/platform/engage_app/delete.py 
* call the new endpoints (POST /apps/) and (DELETE /apps/${app_id})
* (BONUS) remove  https://github.com/Deepomatic/deepocli/blob/a120824498bd62f00489aa71b990c70f6a736706/deepomatic/cli/cmds/platform/engage_app/update.py

To implement the create app-version:
* add new commands create (very similar to https://github.com/Deepomatic/deepocli/blob/a120824498bd62f00489aa71b990c70f6a736706/deepomatic/cli/cmds/platform/engage_app/create.py )
*  reuse the code in https://github.com/Deepomatic/deepocli/blob/a120824498bd62f00489aa71b990c70f6a736706/deepomatic/cli/lib/platform.py#L103 for the create
* call the new endpoints (POST /apps/${app_id}/versions)

### PR DESCRIPTION
**Pull request by oquerne**
close #216

* Use new engage endpoint to create EngageAppVersion
* Not testable as long as engage endpoint is not in production

### CODE REVIEW

File: tests/test_platform.py [hunk 1]
```diff
@@ -103,6 +165,95 @@ def test_appversion(self, no_error_logs):
...
+            and workflow2
+        """
+        with engage_app() as engage_app_id:
+            # Output of create command: New app version 'v<major.minor>' created with id: <id>
+            version1, engage_app_version_id1 = engage_app_version_wrapper(
+                engage_app_id=engage_app_id,
+                workflow=WORKFLOW_PATH,
+                custom_node=CUSTOM_NODES_PATH
+            )
+            assert version1 == "1.0"
```
Comment (maingoh) [inline tests/test_platform.py:206]: `version_number1`

File: tests/test_platform.py [hunk 2]
```diff
@@ -54,11 +60,67 @@ def app_version():
...
+
+    args = "platform engage-app delete --id {}".format(engage_app_id)
+    message = call_deepo(args)
+    assert message == 'Engage App {} deleted'.format(engage_app_id)
+
+
+def engage_app_version_wrapper(engage_app_id,
+                               workflow,
+                               custom_node=None,
+                               from_major=None):
```
Comment (maingoh) [inline tests/test_platform.py:100]: Maybe call it like on the command line and request body

File: tests/test_platform.py [hunk 3]
```diff
@@ -103,6 +165,95 @@ def test_appversion(self, no_error_logs):
...
+
+            with pytest.raises(ClientError) as err:
+                engage_app_version_wrapper(
+                    engage_app_id=engage_app_id,
+                    workflow=WORKFLOW2_PATH,
+                    custom_node=CUSTOM_NODES_PATH,
+                    from_major=version[0]
+                )
+                assert "Bad status code 400" in err
+                assert f"Version already exists v{version}" in err
```
Comment (maingoh) [inline tests/test_platform.py:232]: I am not sure those two lines are executed ? You probably need to unindent them

**Review by maingoh — COMMENTED** [review]
Nice work ! Just a few things to fix and we are good :)

File: deepomatic/cli/lib/platform.py [hunk 4]
```diff
@@ -94,55 +95,85 @@ def __init__(self, client_cls=HTTPHelper):
...
+            response['id']
+        )
+
+    def clone_app_version(self, version_id, recognition_version_ids):
+        data = {'recognition_version_ids': recognition_version_ids}
+
+        response = self.engage_client.post(
+            '{}'.format(
+                self.version_clone_endpoint.format(version_id)
+            ),
```
Comment (ibreschi) [inline deepomatic/cli/lib/platform.py:171]: Like so ? 
```suggestion
            self.version_clone_endpoint.format(version_id),
```

**Review by ibreschi — APPROVED** [review]
It's good for me, maybe fix some of the Hugos comments before merging!

File: tests/test_platform.py [hunk 5]
```diff
@@ -54,11 +60,67 @@ def app_version():
...
+        args += f" -c {custom_node}"
+    if from_major:
+        args += f" --base_major_version {from_major}"
+
+    result = call_deepo(args)
+    _, rhs = result.split(':')
+    engage_app_version_id = rhs.strip()
+
+    assert len(engage_app_version_id) == APP_ID_LEN
+    assert result[0:18] == 'New app version \'v'
```
[the remark this answers is not in the document]
Comment (oquerne) [reply tests/test_platform.py:117]: Well, for this one I'm not sure, I used slicing everywhere else, I thought it was ok to stick with it. Do you have a strong opinion on this ?

[the remark this answers is not in the document]
Comment (maingoh) [reply tests/test_platform.py:117]: No I don't ;) Just find it a bit more readable. Hopefully these slices will go away when we do https://github.com/Deepomatic/deepocli/issues/218

File: deepomatic/cli/lib/platform.py [hunk 6]
```diff
@@ -94,55 +95,85 @@ def __init__(self, client_cls=HTTPHelper):
...
-        with open(workflow_path, 'r') as w:
-            files = {'workflow_yaml': w}
             if custom_nodes_path is not None:
-                with open(custom_nodes_path, 'r') as c:
-                    files['custom_nodes_py'] = c
-                    ret = self.engage_client.post('{}'.format(self.apps_workflow_endpoint),
-                                                  data=data_app, files=files, content_type='multipart/mixed')
+                with open(custom_nodes_path, 'r') as custom_nodes_file:
+                    files['custom_nodes_py'] = custom_nodes_file
+                    response = self.engage_client.post(
```
[the remark this answers is not in the document]
Comment (oquerne) [reply deepomatic/cli/lib/platform.py:145]: nice one

File: deepomatic/cli/lib/site.py [hunk 7]
```diff
@@ -56,7 +56,7 @@ def create(self, name, description, app_version_id):
         if description is not None:
             data['desc'] = description
 
-        ret = self._client.post('/sites', data=data)
```
Comment (oquerne) [inline deepomatic/cli/lib/site.py:59]: To revert

File: deepomatic/cli/cmds/platform/drive_app_version/create.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class CreateCommand(Command):
+    """Create a new DriveAppVersion."""
+
+    def setup(self, subparsers):
+        parser = super(CreateCommand, self).setup(subparsers)
+        parser.add_argument('-n', '--name', required=True, type=str, help="DriveAppVersion name")
+        parser.add_argument('-d', '--description', type=str, help="DriveAppVersion description")
+        parser.add_argument('-a', '--app_id', required=True, type=str, help="DriveApp id for this DriveAppVersion")
```
Comment (maingoh) [inline deepomatic/cli/cmds/platform/drive_app_version/create.py:12]: Maybe here `--drive_app_id` ?

File: deepomatic/cli/cmds/platform/engage_app_version/create.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class CreateCommand(Command):
+    """Create a new EngageAppVersion."""
+
+    def setup(self, subparsers):
+        parser = super(CreateCommand, self).setup(subparsers)
+        parser.add_argument('-w', '--workflow', required=True, type=valid_path, help="Path to the workflow yaml file")
+        parser.add_argument('-c', '--custom_nodes', type=valid_path, help="Path to the custom nodes python file")
+        parser.add_argument('-a', '--app_id', required=True, type=str, help="EngageApp id for this EngageAppVersion")
```
Comment (maingoh) [inline deepomatic/cli/cmds/platform/engage_app_version/create.py:12]: `--engage_app_id` maybe ?

File: deepomatic/cli/cmds/platform/engage_app_version/create_from.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+class CreateFromCommand(Command):
+    """Create an EngageAppVersion from an other one."""
+
+    def setup(self, subparsers):
+        parser = super(CreateFromCommand, self).setup(subparsers)
+        parser.add_argument('--from', dest="origin", required=True, type=str, help="EngageAppVersion id used as base")
+        parser.add_argument('-w', '--workflow', default=None, type=valid_path, help="Path to the workflow yaml file")
+        parser.add_argument('-c', '--custom_nodes', default=None, type=valid_path, help="Path to the custom nodes python file")
+        parser.add_argument('-r', '--recognition-version-ids', nargs="*", type=int,
```
Comment (maingoh) [inline deepomatic/cli/cmds/platform/engage_app_version/create_from.py:13]: nit: you can possibly add the `base_major_version`

File: deepomatic/cli/lib/platform.py [hunk 11]
```diff
@@ -94,55 +90,100 @@ def __init__(self, client_cls=HTTPHelper):
...
+                '{}/{}/versions'.format(self.engage_app_endpoint, app_id),
+                data=data,
+                files=files,
+                content_type='multipart/mixed'
+            )
+        finally:
+            for file in files.values():
+                file.close()
+
+        return "EngageApp version 'v{}.{}' created with id: {}. DriveApp version id: {}".format(
```
Comment (maingoh) [inline deepomatic/cli/lib/platform.py:149]: Nice !

File: deepomatic/cli/lib/platform.py [hunk 12]
```diff
@@ -94,55 +90,100 @@ def __init__(self, client_cls=HTTPHelper):
...
+                files['workflow_yaml'] = open(workflow_path, 'r')
+            if custom_nodes_path:
+                files['custom_nodes_py'] = open(custom_nodes_path, 'r')
+            if files:
+                kwargs['files'] = files
+                kwargs['content_type'] = "multipart/mixed"
+
+            response = self.engage_client.post(
+                self.version_create_from_endpoint.format(origin),
+                **kwargs
```
Comment (maingoh) [inline deepomatic/cli/lib/platform.py:179]: nit: I feel it was easier to just have three variables `data`, `files` and `content_type` and to send them here

File: tests/test_site.py [hunk 13]
```diff
@@ -233,19 +236,20 @@ def test_all(self, no_error_logs):
...
-        with app_version() as (app_version_id, app_id):
-            args = "site create -n test_si -d xyz -v {}".format(app_version_id)
-            result = call_deepo(args)
-            message, site_id = result.split(':')
-            assert message == 'New site created with id'
-
-            args = "site update --id {} --app_version_id {}".format(site_id, app_version_id)
-            message = call_deepo(args)
-            assert message == 'Site{} updated'.format(site_id)
+        with drive_app() as drive_app_id:
```
Comment (maingoh) [inline tests/test_site.py:239]: The test was skipped :O if you can fix it that can be nice !

**Review by maingoh — APPROVED** [review]
https://www.youtube.com/watch?v=fE4n_LUlOTg :partying_face: 

Can you here disambiguate `--drive_app_id`:
https://github.com/Deepomatic/deepocli/blob/c1a19c7f259c5b5449cef4d56d053327c8217fdc/deepomatic/cli/cmds/site/create.py#L14
https://github.com/Deepomatic/deepocli/blob/c1a19c7f259c5b5449cef4d56d053327c8217fdc/deepomatic/cli/cmds/site/update.py#L13

**Pull request merged**